scanner/src/main/kotlin/experimental/FileBasedNestedProvenanceStorage.kt

```diff
@@ -0,0 +1,99 @@
+import java.io.IOException
+
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.yamlMapper
+import org.ossreviewtoolkit.utils.common.collectMessagesAsString
+import org.ossreviewtoolkit.utils.common.fileSystemEncode
+import org.ossreviewtoolkit.utils.core.log
+import org.ossreviewtoolkit.utils.core.showStackTrace
+import org.ossreviewtoolkit.utils.core.storage.FileStorage
+
+class FileBasedNestedProvenanceStorage(private val backend: FileStorage) : NestedProvenanceStorage {
+    override fun readNestedProvenance(root: RepositoryProvenance): NestedProvenanceResolutionResult? =
+        readResults(root).firstOrNull { it.nestedProvenance.root == root }
+
+    private fun readResults(root: RepositoryProvenance): List<NestedProvenanceResolutionResult> {
+        val path = storagePath(root)
+
+        return runCatching {
+            backend.read(path).use { input ->
+                yamlMapper.readValue<List<NestedProvenanceResolutionResult>>(input)
+            }
+        }.getOrElse {
+            when (it) {
+                is FileNotFoundException -> {
+                    // If the file cannot be found it means no scan results have been stored, yet.
+                    emptyList()
+                }
+                else -> {
+                    log.info {
+                        "Could not read resolved nested provenances for '$root' from path '$path': " +
+                                it.collectMessagesAsString()
+                    }
+
+                    emptyList()
+                }
+            }
+        }
+    }
+
+    override fun putNestedProvenance(root: RepositoryProvenance, result: NestedProvenanceResolutionResult) {
+        val results = readResults(root).toMutableList()
+        results.removeAll { it.nestedProvenance.root == root }
+        results += result
+
+        val path = storagePath(root)
+        val yamlBytes = yamlMapper.writeValueAsBytes(results)
+        val input = ByteArrayInputStream(yamlBytes)
+
+        runCatching {
+            backend.write(path, input)
+            log.debug { "Stored resolved nested provenance for '$root' at path '$path'." }
+        }.onFailure {
+            when (it) {
+                is IllegalArgumentException, is IOException -> {
+                    it.showStackTrace()
+
+                    log.warn {
+                        "Could not store resolved nested provenance for '$root' at path '$path': " +
+                                it.collectMessagesAsString()
+                    }
+                }
+                else -> throw it
+            }
+        }
+    }
+}
+
+private const val FILE_NAME = "resolved_nested_provenance.yml"
+
+private fun storagePath(root: RepositoryProvenance) =
+    "${root.vcsInfo.type.toString().fileSystemEncode()}/" +
+            "${root.vcsInfo.url.fileSystemEncode()}/" +
+            "${root.resolvedRevision.fileSystemEncode()}/" +
+            FILE_NAME
```

Comment:
Just an idea: Maybe use something like `listOf().joinToString("/") { it.fileSystemEncode() }`?

Reply:
I tried it but kept the current one, it's a bit repetitive but also more explicit.